ecs-cli/modules/cli/local/network/teardown.go

```diff
@@ -14,25 +14,134 @@
+// These functions can be used together to remove a network once unwanted containers are stopped.
 type LocalEndpointsStopper interface {
+	networkInspector
+	containerStopper
+	containerRemover
+	networkRemover
+}
+
+type networkInspector interface {
 	NetworkInspect(ctx context.Context, networkID string, options types.NetworkInspectOptions) (types.NetworkResource, error)
+}
+
+type containerStopper interface {
 	ContainerStop(ctx context.Context, containerID string, timeout *time.Duration) error
+}
+
+type containerRemover interface {
+	ContainerRemove(ctx context.Context, containerID string, options types.ContainerRemoveOptions) error
+}
+
+type networkRemover interface {
 	NetworkRemove(ctx context.Context, networkID string) error
 }
 
-// Teardown stops the Local Endpoints container and removes the Local network created by Setup.
+// Teardown removes both the Local Endpoints container and the Local network created by Setup.
 // If there are other containers running in the network besides the endpoints container, this function does nothing.
 //
 // If there is any unexpected errors, we exit the program with a fatal log.
-func Teardown(dockerClient *LocalEndpointsStopper) {
+func Teardown(dockerClient LocalEndpointsStopper) {
+	if hasRunningTasksInNetwork(dockerClient) {
+		return
+	}
+	logrus.Infof("The network %s has no more running tasks, stopping the endpoints containers...", EcsLocalNetworkName)
+
+	stopEndpointsContainer(dockerClient)
+	removeEndpointsContainer(dockerClient)
+	removeLocalNetwork(dockerClient)
+}
+
+// hasRunningTasksInNetwork returns true if there are other containers besides the
+// endpoints container running in the local network, false otherwise.
+func hasRunningTasksInNetwork(d networkInspector) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), dockerTimeout)
+	defer cancel()
+
+	resp, err := d.NetworkInspect(ctx, EcsLocalNetworkName, types.NetworkInspectOptions{})
+	if err != nil {
+		logrus.Fatalf("Failed to inspect network %s due to %v", EcsLocalNetworkName, err)
+	}
+
+	if len(resp.Containers) > 1 {
+		// Has other containers running in the network
+		return true
+	}
+
+	for _, container := range resp.Containers {
+		if container.Name != localEndpointsContainerName {
+			// The only container running in the network is a task without the endpoints container.
+			// This scenario should not happen unless the user themselves stopped the endpoints container.
+			logrus.Warnf("The %s container is running in the %s network without the %s container, please stop it first",
+				container.Name, EcsLocalNetworkName, localEndpointsContainerName)
+			return true
+		}
+	}
+
+	return false
+}
+
+func stopEndpointsContainer(d containerStopper) {
+	ctx, cancel := context.WithTimeout(context.Background(), dockerTimeout)
+	defer cancel()
+
+	err := d.ContainerStop(ctx, localEndpointsContainerName, nil)
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), "no such container") {
```

Comment:
Why is this needed?

Reply:
Great question, looks like the comment below wasn't precise enough :P 

This error happens if the user stopped all their task containers with `docker stop` instead of `ecs-cli local stop`. In that situation, `hasRunningTasksInNetwork()` will return `false` and we will try to stop the endpoints container again. But since it was already stopped this error is returned. 

We don't want to `log.Fatal` while trying to stop an already stopped container, it should be a no-op hence this if-statement. 

I updated the comment to `// The containers in the network were already stopped by the user using "docker stop", do nothing.`

Reply:
An alternative is to not invoke `stopEndpointsContainer()` or `removeEndpointsContainer()` if there are 0 containers in the local network.

However, I think that approach pushes complexity upwards instead of downwards (into the function) so I don't prefer it as much. Also if you run `ecs-cli local stop` on the same task concurrently (two terminals for example) then one of those could trigger this error and fail with fatal instead of continuing cleanly by not doing any work.

Reply:
Got it... my only concern is that this is yet another place where we are using string matching on errors returned by Docker :/

Reply:
Yeah, I agree but I think Docker locked themselves and can't change their error messages otherwise they'll break everyone's code :P  

Here is a similar example from the agent: https://github.com/aws/amazon-ecs-agent/blob/24c71162368b3a9090689532e5c035fe2bccf133/agent/dockerclient/dockerapi/docker_client.go#L693

Reply:
This is my least favorite thing in Go...